mmv1/third_party/terraform/website/docs/r/bigtable_instance.html.markdown

```diff
@@ -131,11 +131,7 @@ for a `DEVELOPMENT` instance.
 
 * `kms_key_name` - (Optional) Describes the Cloud KMS encryption key that will be used to protect the destination Bigtable cluster. The requirements for this key are: 1) The Cloud Bigtable service account associated with the project that contains this cluster must be granted the `cloudkms.cryptoKeyEncrypterDecrypter` role on the CMEK key. 2) Only regional keys can be used and the region of the CMEK key must match the region of the cluster.
 
-!> **Warning**: Modifying this field will cause Terraform to delete/recreate the entire resource. 
-
--> **Note**: To remove this field once it is set, set the value to an empty string. Removing the field entirely from the config will cause the provider to default to the backend value.
```

Comment:
I believe this statement to still apply since the value is computed. I tested it with a local deployment. Could this statement be retained?

Reply:
The value is computed. If you remove the field from any given cluster the value will be unset, but will not force a recreation. We cannot remove this computed property, it would be a breaking change. The guidance to set to an empty string allows the user to force recreation while unsetting the kms property. You're saying setting an kms key to an empty string will currently cause an error? Is that correct?

Reply:
> The value is computed. What computed means is the value is always read from the api if not defined. If you remove the `kms` field from any given cluster it will not recognize any difference, thus will not force a recreation.
> 

This behavior is understood, we don't need that note.

> We cannot remove this computed property, it would be a breaking change. The guidance to set to an empty string allows the user to force recreation while unsetting the kms property. You're saying setting an kms key to an empty string will currently cause an error? Is that correct?

This PR doesn't remove the "computed" property.

Setting to an empty string to a kms key in a cluster will be an error in our API, hence TF will try to recreate the instance instead.

My point is the note we currently have is more confusing than being helpful. We should remove it.

Reply:
> This behavior is understood, we don't need that note.
I am simply providing context. I am unaware of your personal experience and not insulting your knowledge.

> This PR doesn't remove the "computed" property.
I didn't suggest that it does. Computed is harmful in some scenarios. The alternative of removing it, while desirable, wouldn't be possible. 

> My point is the note we currently have is more confusing than being helpful. We should remove it.
I've tested the deployment locally with an empty string. It was able to unset the kms key value by using an empty string. We should retain the note explaining how to unset it.

Reply:
> > This behavior is understood, we don't need that note.
> 
> I am simply providing context. I am unaware of your personal experience and not insulting your knowledge.

No, I didn't take it that way. I only wished to ack that I am aware of it. Thanks for your context!

> 
> > This PR doesn't remove the "computed" property.
> 
> I didn't suggest that it does. Computed is harmful in some scenarios. The alternative of removing it, while desirable, wouldn't be possible.
> 
> > My point is the note we currently have is more confusing than being helpful. We should remove it.
> 
> I've tested the deployment locally with an empty string. It was able to unset the kms key value by using an empty string. We should retain the note explaining how to unset it.

Let's focus on this sentence "To remove this field once it is set, set the value to an empty string."
This seems to suggest that a user can set the value of `kms_key_name` to an empty string after it is set to a key.
I know rekey is not supported, meaning the key is not updatable, hence ForceNew.

I tested locally:
1. Create an instance/cluster with `kms_key_name` set to "x".
2. Set `kms_key_name` to an empty string.

The second step is not allowed unless the instance is recreated.

Therefore, I believe "To remove this field once it is set, set the value to an empty string." is confusing.

Scott, let me know how you tested it. Thanks!

Reply:
Correct, it was through recreation. My concern is that this recreation can be surprising. My intention is to retain the knowledge of what will happen on removal of the field. The developer does so with an intention so if that intention is subverted then that experience can be surprising.  Here are the options as I see it.

* Keep the note as is. As you mention, the note could be confusing as stated, suggesting the field can be changed.
* Rephrase the note to be encapsulate the behavior better.
* Remove the note.

Removing the note makes me nervous since there is a loss of information. The developer's original intention was to provide the experience to the user so they would know what to expect. 

```markdown
-> **Note**: Removing the field `kms_key_name` entirely from the config will cause the provider to default to the backend value. Destroy and recreate the resource `terraform destroy -target <resource>` to actuate this removal.
```

Reply:
> Correct, it was through recreation. My concern is that this recreation can be surprising. My intention is to retain the knowledge of what will happen on removal of the field. The developer does so with an intention so if that intention is subverted then that experience can be surprising. Here are the options as I see it.
> 
> * Keep the note as is. As you mention, the note could be confusing as stated, suggesting the field can be changed.
> * Rephrase the note to be encapsulate the behavior better.
> * Remove the note.
> 
> Removing the note makes me nervous since there is a loss of information. The developer's original intention was to provide the experience to the user so they would know what to expect.
> 
> ```gfm
> -> **Note**: Removing the field `kms_key_name` entirely from the config will cause the provider to default to the backend value. Destroy `terraform destroy -target <resource>`  and recreate the resource to actuate this removal.
> ```

Sounds good. I added this part back:  "Removing the field entirely from the config will cause the provider to default to the backend value." I thought this behavior is typical.